A39-xds-http-filters.md

```diff
@@ -0,0 +1,152 @@
+API that third parties can use to implement their own filters, although
+that may come later.  For now, we are focusing on creating a structure
+that will enable us to implement functionality that, in Envoy, is
+provided by its out-of-the-box filters (e.g., fault injection or
+authorization), which we also want to be provided out-of-the-box in gRPC.
+
+### Related Proposals: 
+
+- [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+- [A30: xDS v3 Support](https://github.com/grpc/proposal/blob/master/A30-xds-v3.md)
+- [A31: gRPC xDS Timeout Support and Config Selector Design](https://github.com/grpc/proposal/blob/master/A31-xds-timeout-support-and-config-selector.md)
+
+## Proposal
+
+In the gRPC client channel, the xDS HTTP filters will run in between
+name resolution and load balancing.  The set of filters to run will be
+determined by the `ConfigSelector` returned by the resolver, which is
+where gRPC implements the xDS routing functionality.
+
+In C-core, the xDS HTTP filters will be implemented as C-core filters.
+In Java and Go, they will be implemented as interceptors.
+
+### Limitations
+
+gRPC will support HTTP filters in xDS v3 only.  Filters will continue to
+be ignored when speaking xDS v2.  This will avoid the need to support
+both v2 and v3 type names in filter config message types.
+
+In Envoy, HTTP filters have access to an API to tell Envoy to recompute
+the route, which they can use after modifying request state that can
+affect the choice of route (e.g., changing a header).  Initially, gRPC will
+not support any such mechanism, although this is something that we may
+add later, if/when it becomes necessary.
+
+### xDS API Fields
+
+gRPC will add support for the following fields in the xDS APIs:
+- [`HttpConnectionManager.http_filters`](https://github.com/envoyproxy/envoy/blob/2ee9543fbb0fccdb69d4dcd8cb57a19743afb94b/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L273)
+- [`VirtualHost.typed_per_filter_config`](https://github.com/envoyproxy/envoy/blob/2ee9543fbb0fccdb69d4dcd8cb57a19743afb94b/api/envoy/config/route/v3/route_components.proto#L142)
+- [`Route.typed_per_filter_config`](https://github.com/envoyproxy/envoy/blob/2ee9543fbb0fccdb69d4dcd8cb57a19743afb94b/api/envoy/config/route/v3/route_components.proto#L252)
+- [`WeightedCluster.ClusterWeight.typed_per_filter_config`](https://github.com/envoyproxy/envoy/blob/2ee9543fbb0fccdb69d4dcd8cb57a19743afb94b/api/envoy/config/route/v3/route_components.proto#L366)
+
+The `HttpConnectionManager.http_filters` field is an ordered list of
+filters.  In each entry, gRPC will look at the following fields:
+- [`name`](https://github.com/envoyproxy/envoy/blob/2ee9543fbb0fccdb69d4dcd8cb57a19743afb94b/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L833):
```

Comment:
Should we NACK if the name is the empty string, or is this a valid setting as long as only one filter entry has it?

Reply:
Ah, you're right - I missed that.